File: Data Structures & Algorithms/permutation-string/submission-0.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        if(s2.length() < s1.length())
            return false;

        int n = s1.length();
        int[] cnt = new int[26];
        for(int i=0; i<n; i++){
            cnt[s1.charAt(i)-'a']++;
        }

        int matches = 0;
        for(int i=0; i<26; i++){
            if(cnt[i]==0)
                matches++;
        }
        for(int i=0; i<s2.length(); i++){
            int in = s2.charAt(i)-'a';

            cnt[in]--;

            if(cnt[in]==0)
                matches++;
            
            if(i>=n){
                int out = s2.charAt(i-n)-'a';
                if(cnt[out]==0)
                    matches--;
                cnt[out]++;                
            }

            if(matches == 26)
                return true;
        }

        return false;
    }
}

File: Data Structures & Algorithms/is-anagram/submission-0.cs
public class Solution {
    public bool IsAnagram(string s, string t) {
        if(s.Length != t.Length){
            return false;
        }
        char[] c1 = new char[26];
        char[] c2 = new char[26];

        for(int i=0; i<s.Length; i++){
            c1[s[i]-'a']++;
            c2[t[i]-'a']++;
        }

        for(int i=0; i<26; i++){
            if(c1[i] != c2[i])
                return false;
        }

        return true;
    }
}
